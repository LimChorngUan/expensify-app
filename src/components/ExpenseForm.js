import React, { useState } from 'react'

export const FORM_TYPE = {
  ADD: `add`,
  EDIT: `edit`,
}

const BUTTON_TEXT = {
  [FORM_TYPE.ADD]: `Add Expense`,
  [FORM_TYPE.EDIT]: `Edit Expense`,
}

const ExpenseForm = ({
  handleSubmit,
  expense = {},
  formType
}) => {
  const [description, setDescription] = useState(expense.description || ``)
  const [amount, setAmount] = useState(expense.amount || 0)
  const [createdAt, setCreatedAt] = useState(expense.createdAt || 0)
  const [note, setNote] = useState(expense.note || ``)

  return (
    <div>
      <form
        onSubmit={(e) => {
          e.preventDefault()
          handleSubmit({
            description,
            amount,
            createdAt,
            note,
          })
        }}
        data-test-id="expense-form"
      >
        <input
          type="text"
          placeholder="description"
          value={description}
          onChange={(e) => setDescription(e.target.value)}
          data-test-id="description-input"
        />
        <input
          type="number"
          placeholder="amount"
          value={amount}
          onChange={(e) => setAmount(e.target.value)}
          data-test-id="amount-input"
        />
        <input
          type="number"
          placeholder="created at"
          value={createdAt}
          onChange={(e) => setCreatedAt(e.target.value)}
        />
        <textarea
          placeholder="Write your notes here (optional)"
          value={note}
          onChange={(e) => setNote(e.target.value)}
          data-test-id="note-input"
        />
        <button type="submit">{BUTTON_TEXT[formType]}</button>
      </form>
    </div>
  )
}

export default ExpenseForm