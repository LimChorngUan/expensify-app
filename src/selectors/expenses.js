import _ from 'lodash'

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate}) => {
  const filteredExpenses = expenses.filter(expense => {
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
    const startDateMatch = typeof startDate !== `number` || expense.createdAt >= startDate
    const endDateMatch = typeof endDate !== `number` || expense.createdAt <= endDate

    return textMatch && startDateMatch && endDateMatch
  })

  if (sortBy === `amount`) {
    return _.sortBy(filteredExpenses, [`amount`])
  } else if (sortBy === `date`) {
    return _.sortBy(filteredExpenses, [`createdAt`])
  }
}

export default getVisibleExpenses