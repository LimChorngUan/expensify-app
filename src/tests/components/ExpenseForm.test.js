import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import ExpenseForm, { FORM_TYPE } from '../../components/ExpenseForm'
import { expenses } from '../fixture/expenses'
import { generateTestId } from '../utils'

describe(`Component: ExpenseForm`, () => {
  describe(`should render the correct form type`, () => {
    test(`Add expense form type`, () => {
      const wrapper = shallow(<ExpenseForm formType={FORM_TYPE.ADD} />)
      expect(toJson(wrapper)).toMatchSnapshot()
    })

    test(`Edit expense form type`, () => {
      const wrapper = shallow(<ExpenseForm formType={FORM_TYPE.EDIT} />)
      expect(toJson(wrapper)).toMatchSnapshot()
    })
  })

  describe(`should render correctly with provided expense`, () => {
    const expense = expenses[0]

    test(`Add expense form`, () => {
      const wrapper = shallow(<ExpenseForm formType={FORM_TYPE.ADD} expense={expense} />)
      expect(toJson(wrapper)).toMatchSnapshot()
    })

    test(`Edit expense form`, () => {
      const wrapper = shallow(<ExpenseForm formType={FORM_TYPE.EDIT} expense={expense} />)
      expect(toJson(wrapper)).toMatchSnapshot()
    })
  })

  describe(`Test user iteraction`, () => {
    let component
    let mockedEvent

    beforeEach(() => {
      component = mount(
        <ExpenseForm
          formType={FORM_TYPE.ADD}
          handleSubmit={jest.fn()}
        />
        )
    })

    test(`description input value is updated on input change`, () => {
      mockedEvent = {
        target: {
          value: `A`,
        },
      }

      component.find(generateTestId(`description-input`)).simulate(`change`, mockedEvent)
      expect(component.find(generateTestId(`description-input`)).props().value).toEqual(`A`)
    })

    test(`note text area value is updated on text area change`, () => {
      mockedEvent = {
        target: {
          value: `aaaaa`,
        },
      }

      component.find(generateTestId(`note-input`)).simulate(`change`, mockedEvent)
      expect(component.find(generateTestId(`note-input`)).props().value).toEqual(`aaaaa`)
    })

    test(`amount input value is updated on input change`, () => {
      mockedEvent = {
        target: {
          value: 20,
        },
      }

      component.find(generateTestId(`amount-input`)).simulate(`change`, mockedEvent)
      expect(component.find(generateTestId(`amount-input`)).props().value).toEqual(20)
    })

    test(`handleSubmit is called on form submit`, () => {
      mockedEvent = { preventDefault: jest.fn() }

      component.find(generateTestId(`expense-form`)).simulate(`submit`, mockedEvent)
      expect(component.props().handleSubmit).toBeCalledTimes(1)
    })

    test(`handleSubmit is called with correct argument`, () => {
      component.find(generateTestId(`description-input`)).simulate(`change`, { target: { value: `A` }})
      component.find(generateTestId(`amount-input`)).simulate(`change`, { target: { value: 20 }})
      component.find(generateTestId(`note-input`)).simulate(`change`, { target: { value: `aaaaa` }})

      component.find(generateTestId(`expense-form`)).simulate(`submit`, { preventDefault: jest.fn()})
      expect(component.props().handleSubmit).toBeCalledWith({
        description: `A`,
        amount: 20,
        createdAt: 0,
        note: `aaaaa`,
      })
    })
  })
})