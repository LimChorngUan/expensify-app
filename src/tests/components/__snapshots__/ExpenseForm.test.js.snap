// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ExpenseForm should render correctly with provided expense Add expense form 1`] = `
<div>
  <form
    data-test-id="expense-form"
    onSubmit={[Function]}
  >
    <input
      data-test-id="description-input"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value="A"
    />
    <input
      data-test-id="amount-input"
      onChange={[Function]}
      placeholder="amount"
      type="number"
      value={1}
    />
    <input
      onChange={[Function]}
      placeholder="created at"
      type="number"
      value={1}
    />
    <textarea
      data-test-id="note-input"
      onChange={[Function]}
      placeholder="Write your notes here (optional)"
      value="aaa"
    />
    <button
      type="submit"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`Component: ExpenseForm should render correctly with provided expense Edit expense form 1`] = `
<div>
  <form
    data-test-id="expense-form"
    onSubmit={[Function]}
  >
    <input
      data-test-id="description-input"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value="A"
    />
    <input
      data-test-id="amount-input"
      onChange={[Function]}
      placeholder="amount"
      type="number"
      value={1}
    />
    <input
      onChange={[Function]}
      placeholder="created at"
      type="number"
      value={1}
    />
    <textarea
      data-test-id="note-input"
      onChange={[Function]}
      placeholder="Write your notes here (optional)"
      value="aaa"
    />
    <button
      type="submit"
    >
      Edit Expense
    </button>
  </form>
</div>
`;

exports[`Component: ExpenseForm should render the correct form type Add expense form type 1`] = `
<div>
  <form
    data-test-id="expense-form"
    onSubmit={[Function]}
  >
    <input
      data-test-id="description-input"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value=""
    />
    <input
      data-test-id="amount-input"
      onChange={[Function]}
      placeholder="amount"
      type="number"
      value={0}
    />
    <input
      onChange={[Function]}
      placeholder="created at"
      type="number"
      value={0}
    />
    <textarea
      data-test-id="note-input"
      onChange={[Function]}
      placeholder="Write your notes here (optional)"
      value=""
    />
    <button
      type="submit"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`Component: ExpenseForm should render the correct form type Edit expense form type 1`] = `
<div>
  <form
    data-test-id="expense-form"
    onSubmit={[Function]}
  >
    <input
      data-test-id="description-input"
      onChange={[Function]}
      placeholder="description"
      type="text"
      value=""
    />
    <input
      data-test-id="amount-input"
      onChange={[Function]}
      placeholder="amount"
      type="number"
      value={0}
    />
    <input
      onChange={[Function]}
      placeholder="created at"
      type="number"
      value={0}
    />
    <textarea
      data-test-id="note-input"
      onChange={[Function]}
      placeholder="Write your notes here (optional)"
      value=""
    />
    <button
      type="submit"
    >
      Edit Expense
    </button>
  </form>
</div>
`;
